# Cleaning and analyzing eBay Car Sales Data

**objective**: analysis of eBay Car Sales
**Dataset**: eBay Kleinanzeigen, a classifieds section of the German eBay website

** Data dictionary**
* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which which year the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which which year the car was * first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.


In [1]:
import pandas as pd

In [2]:
autos = pd.read_csv("autos.csv", encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

## Adjusting and rewording column names

In [5]:
column_names = autos.columns
print(column_names)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [6]:
autos.rename({
    "yearOfRegistration" : "registration_year",
    "monthOfRegistration" : "registration_month",
    "notRepairedDamage" : "unrepaired_damage",
    "dateCreated" : "ad_created",
    'dateCrawled' : 'date_crawled',
    'offerType' : 'offer_type',
    'vehicleType' : 'vehicle_type',
    'fuelType' : 'fuel_type',
    'nrOfPictures' : 'nr_of_pictures',
    'postalCode' : 'postal_code',
    'lastSeen' : 'last_seen'},
    axis=1,
    inplace=True)

In [7]:
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


## Let's look for redundant or unnecessary columns

In [8]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

**Observations**
* 'seller' is systematically "privat" but for one entry
* 'offer_type' is systematically "angebot" but for one entry
* 'nf_of_pictures' is systematically 0

**Conclusion**
* We'll drop all 3 columns

In [9]:
autos.drop(labels=["offer_type", "nr_of_pictures", "seller"], axis=1, inplace=True)

## Let's convert to numeric values where relevant

In [10]:
autos["price"] = (autos["price"]
                  .str.replace("$","")
                  .str.replace(",","")
                  .astype(int))

autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [11]:
autos['odometer'] = (autos['odometer']
                     .str.replace("km","")
                     .str.replace(",","")
                     .astype(int))
autos.rename({'odometer' : 'odometer_km'}, axis=1, inplace=True)
autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

## Let's look for inconsistent data

In [12]:
print(autos[['odometer_km', 'price']].describe())
print(autos['odometer_km'].value_counts())
print(autos['price'].value_counts())

         odometer_km         price
count   50000.000000  5.000000e+04
mean   125732.700000  9.840044e+03
std     40042.211706  4.811044e+05
min      5000.000000  0.000000e+00
25%    125000.000000  1.100000e+03
50%    150000.000000  2.950000e+03
75%    150000.000000  7.200000e+03
max    150000.000000  1.000000e+08
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64
0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64


**Observations**
* 1421 cars which price is zero.
* A maximum car price of '$100,000,000'. 

**Questions**
* How many cars have unrealistically low price?
* How many cars have unrealistically high prices ?

In [13]:
autos['price'].value_counts().head()

0       1421
500      781
1500     734
2500     643
1000     639
Name: price, dtype: int64

In [14]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

**Conclusions**
* Mileage information looks consistent. unsurprisingly, cars sold on eBay have a higher mileage in general
* We'll consider the zero price cars as outliers, and remove them
* As well as the cars with a price above '$500,000'

In [15]:
autos = autos[(autos['price'] > 0) & (autos['price'] < 500000)]

In [16]:
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

# Cleaning and analysis of date informations
We focus on the date fields of the data set
* date_crawled
* last_seen
* ad_created
* registration_month
* registration_year

**Objective**
We want to analyze the date range covered by the data set

In [17]:
autos[['date_crawled', 'last_seen', 'ad_created', 'registration_month', 'registration_year']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48565 entries, 0 to 49999
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        48565 non-null  object
 1   last_seen           48565 non-null  object
 2   ad_created          48565 non-null  object
 3   registration_month  48565 non-null  int64 
 4   registration_year   48565 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 2.2+ MB


**Observations**
* date_crawled, last_seen, ad_created are stored as string objects. We will need to convert them for data range analysis
* registration_month and registration_year are stored as numerical values which allows analysis

Let's start preparing the first three columns for analysis

In [18]:
autos[['date_crawled', 'last_seen', 'ad_created']].head()

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

**Observations**
* The date is stored in the first 10 characters of the string for each column
* The remainder of the string indicates a time of day, which we don't need for the moment

**Conclusion**
* Let's focus on the 10 first characters for each of the strings
* We don't want to drop the rest of the information, so we start by duplicating the data of interest

In [19]:
crawled_dates = autos['date_crawled'].str[:10]
created_dates =  autos['ad_created'].str[:10]
last_seen_dates =  autos['last_seen'].str[:10]

*Analysis of date_crawled*

In [20]:
crawled_dates.value_counts(normalize=True, dropna=False)

2016-04-03    0.038608
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-12    0.036920
2016-03-14    0.036549
2016-04-04    0.036487
2016-03-07    0.036014
2016-04-02    0.035478
2016-03-28    0.034860
2016-03-19    0.034778
2016-03-15    0.034284
2016-03-29    0.034099
2016-03-30    0.033687
2016-04-01    0.033687
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-22    0.032987
2016-03-11    0.032575
2016-03-23    0.032225
2016-03-26    0.032204
2016-03-10    0.032184
2016-03-31    0.031834
2016-03-17    0.031628
2016-03-25    0.031607
2016-03-27    0.031092
2016-03-16    0.029610
2016-03-24    0.029342
2016-03-05    0.025327
2016-03-13    0.015670
2016-03-06    0.014043
2016-04-05    0.013096
2016-03-18    0.012911
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [21]:
(crawled_dates
 .value_counts(normalize=True, dropna=False)
 .sort_index())

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

**Observations**
* Ads were crawled daily and uniformally over the perios, but for 6 dates with lower frequency
* date_crawled ranges from March 5th trough April 7th

*Analysis of ad_created*

In [22]:
created_dates.value_counts(normalize=True, dropna=False)

2016-04-03    0.038855
2016-03-20    0.037949
2016-03-21    0.037579
2016-04-04    0.036858
2016-03-12    0.036755
                ...   
2016-01-07    0.000021
2016-02-11    0.000021
2016-02-08    0.000021
2015-12-30    0.000021
2016-02-09    0.000021
Name: ad_created, Length: 76, dtype: float64

In [23]:
(created_dates
 .value_counts(dropna=False)
 .sort_index()
 .head(20))

2015-06-11    1
2015-08-10    1
2015-09-09    1
2015-11-10    1
2015-12-05    1
2015-12-30    1
2016-01-03    1
2016-01-07    1
2016-01-10    2
2016-01-13    1
2016-01-14    1
2016-01-16    1
2016-01-22    1
2016-01-27    3
2016-01-29    1
2016-02-01    1
2016-02-02    2
2016-02-05    2
2016-02-07    1
2016-02-08    1
Name: ad_created, dtype: int64

In [24]:
(created_dates
 .value_counts(dropna=False)
 .sort_index()
 .tail(40))

2016-02-28      10
2016-02-29       8
2016-03-01       5
2016-03-02       5
2016-03-03      42
2016-03-04      72
2016-03-05    1112
2016-03-06     744
2016-03-07    1687
2016-03-08    1618
2016-03-09    1610
2016-03-10    1549
2016-03-11    1598
2016-03-12    1785
2016-03-13     826
2016-03-14    1709
2016-03-15    1652
2016-03-16    1463
2016-03-17    1519
2016-03-18     660
2016-03-19    1636
2016-03-20    1843
2016-03-21    1825
2016-03-22    1593
2016-03-23    1557
2016-03-24    1422
2016-03-25    1542
2016-03-26    1567
2016-03-27    1505
2016-03-28    1699
2016-03-29    1653
2016-03-30    1627
2016-03-31    1548
2016-04-01    1636
2016-04-02    1707
2016-04-03    1887
2016-04-04    1790
2016-04-05     574
2016-04-06     158
2016-04-07      61
Name: ad_created, dtype: int64

**Observations**
* Ads were created from June 11th, 2015 through April 7th, 2016
* While older ads are very few as the items are not on sale anymore, but exceptionally, the latest ads are less. Possible explanation: it may take a few days for an author to complete the information for an ad.
* The bulk of the creation dates start from March 5th, 2016

*Analysis of last_seen dates*

In [25]:
last_seen_dates.value_counts(normalize=True, dropna=False)

2016-04-06    0.221806
2016-04-07    0.131947
2016-04-05    0.124761
2016-03-17    0.028086
2016-04-03    0.025203
2016-04-02    0.024915
2016-03-30    0.024771
2016-04-04    0.024483
2016-03-12    0.023783
2016-03-31    0.023783
2016-04-01    0.022794
2016-03-29    0.022341
2016-03-22    0.021373
2016-03-28    0.020859
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-23    0.018532
2016-03-26    0.016802
2016-03-16    0.016452
2016-03-15    0.015876
2016-03-19    0.015834
2016-03-27    0.015649
2016-03-14    0.012602
2016-03-11    0.012375
2016-03-10    0.010666
2016-03-09    0.009595
2016-03-13    0.008895
2016-03-08    0.007413
2016-03-18    0.007351
2016-03-07    0.005395
2016-03-06    0.004324
2016-03-05    0.001071
Name: last_seen, dtype: float64

In [26]:
(last_seen_dates
 .value_counts(dropna=False)
 .sort_index())

2016-03-05       52
2016-03-06      210
2016-03-07      262
2016-03-08      360
2016-03-09      466
2016-03-10      518
2016-03-11      601
2016-03-12     1155
2016-03-13      432
2016-03-14      612
2016-03-15      771
2016-03-16      799
2016-03-17     1364
2016-03-18      357
2016-03-19      769
2016-03-20     1003
2016-03-21     1002
2016-03-22     1038
2016-03-23      900
2016-03-24      960
2016-03-25      933
2016-03-26      816
2016-03-27      760
2016-03-28     1013
2016-03-29     1085
2016-03-30     1203
2016-03-31     1155
2016-04-01     1107
2016-04-02     1210
2016-04-03     1224
2016-04-04     1189
2016-04-05     6059
2016-04-06    10772
2016-04-07     6408
Name: last_seen, dtype: int64

**Observations**
* last seen date of an ad can be assumed to be the date the ad was removed
* Almost half the ads were last seen by the crawler in the last 3 days of the period. The crawler may not crawl through all ads every day, which explains why a bigger amount of ads were last seen on the past 3 days
* As one goes back in time, there are, overall, less ads that were last seen on a given date

**Recap on date ranges observations**
* Ads were created from June 11th, 2015 through April 7th, 2016
* The bulk of the creation dates start from March 5th, 2016
* Ads were crawled from March 5th trough April 7th, 2016
* Ads were last seen from March 5th, 2016, with half of them in the latest 3 days of April 2016

**Registration year analysis**

In [27]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

**Observations**
* Registration years concentrate in a period of 10 years around 2004
* We observe unrealistic registration years whether in the past or in the future, which requires cleaning


## Cleaning for unrealistic registration dates

In [28]:
(autos['registration_year']
 .value_counts(dropna=False)
 .sort_index()
 .head(20))

1000    1
1001    1
1111    1
1800    2
1910    5
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
1950    3
1951    2
1952    1
1953    1
1954    2
Name: registration_year, dtype: int64

In [29]:
(autos['registration_year']
 .value_counts(dropna=False)
 .sort_index()
 .tail(20))

2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, dtype: int64

**Observations**
* We confirm the presence of inconsistent -too old or future- registrations. There are less than 2000 such records, which is less than 4% of our records. Thus we shouldn't be losing significant data

**Required corrections**
* Obviously, we should discard ads with registration years before 1900
* We should also discard registrations which are after the ads date range, that is 2017 registrations and beyond

In [30]:
autos = autos[autos['registration_year'].between(1900,2016)]

(autos['registration_year']
 .value_counts(normalize=True)
 .head(20))

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
2008    0.047450
2009    0.044665
1997    0.041794
2011    0.034768
2010    0.034040
1996    0.029412
2012    0.028063
1995    0.026285
2016    0.026135
2013    0.017202
Name: registration_year, dtype: float64

In [31]:
(autos['registration_year']
 .value_counts(normalize=True)
 .tail(20))

1956    0.000086
1958    0.000086
1937    0.000086
1962    0.000086
1950    0.000064
1954    0.000043
1941    0.000043
1951    0.000043
1934    0.000043
1957    0.000043
1955    0.000043
1953    0.000021
1943    0.000021
1929    0.000021
1939    0.000021
1938    0.000021
1948    0.000021
1927    0.000021
1931    0.000021
1952    0.000021
Name: registration_year, dtype: float64

**Observations**
* The bulk of cars were registered in the late 1990s, with a concentration in the 1999 - 2005 period
* More recent cars, registered in 2010 and after, are less, as we are dealing with second hand car market
* There is a long tail of older registrations which goes back to the 1930s

## Analysis by brand, aggregation ##

*Which brands should we focus on?*

In [32]:
brand_count = (autos["brand"]
 .value_counts(normalize=True, dropna=False)
 .sort_values(ascending=False)
)

In [33]:
top_brands = brand_count[brand_count > 0.05].index
print(top_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


**Observations**
* 6 brands have a market share of 5% or more, we'lll focus on those

In [34]:
mean_price_dict = {}
for brand in top_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    mean_price_dict[brand] = int(mean_price)


In [35]:
print(mean_price_dict)

{'volkswagen': 5402, 'bmw': 8332, 'opel': 2975, 'mercedes_benz': 8628, 'audi': 9336, 'ford': 3749}


**Observations**
* Audi is the most expensive brand, while Opel is the cheapest, among the most sold cars in the german second hand market represented by eBay
* The Ausi mean price is 3 times the Opel mean price
* Mercedes and BMW are the 2 other expensive brands, Ford is the second cheapest brand, Volkswagen is the mid range brand

## Cross-analysis of price and mileage 

*We first build the relevant mileage data

*We first build the relevant pandas data structure

In [36]:
mean_mileage_dict = {}
for brand in top_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    mean_mileage_dict[brand] = int(mean_mileage)


In [37]:
print(mean_mileage_dict)

{'volkswagen': 128707, 'bmw': 132572, 'opel': 129310, 'mercedes_benz': 130788, 'audi': 129157, 'ford': 124266}


*Create series objects out of the 2 data dictionnaries*

In [38]:
mean_price_series = pd.Series(mean_price_dict)

In [39]:
mean_mileage_series = pd.Series(mean_mileage_dict)

*Create a dataframe with the first series (mean price)*

In [40]:
cross_analysis_df = pd.DataFrame(mean_price_series, columns=['mean_price'])

*Add the second series (mean mileage) to the dataframe*

In [41]:
cross_analysis_df['mean_mileage'] = mean_mileage_series

In [42]:
(cross_analysis_df
 .sort_values('mean_price')
 .head(6))

mean_price  mean_mileage
opel                 2975        129310
ford                 3749        124266
volkswagen           5402        128707
bmw                  8332        132572
mercedes_benz        8628        130788
audi                 9336        129157

**Observations**
* Mean mileages are close across the 6 most sold brands in the second hand eBay market
* They differ by 8000km at most, that is 6% or less
* Mean prices differ much more with up to x3 differences 
* The lost mean mileage brand is actually the second cheapest, and the most expensive brand has average mean mileage

**Conclusion**
* Mileage has no significant impact on brand mean price, though a small correlation may be observed